src/clickhouse_fetcher.mjs

```diff
@@ -0,0 +1,114 @@
+    const pastDate = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate() - 28);
+    /* eslint-disable no-unmodified-loop-condition */
+    while (pastDate <= currentDate) {
+        const queryString = `INSERT INTO ${TMP_TABLE_NAME} (pixel, params)
+            WITH extractURLParameters(request) AS params
+            SELECT any(pixel), arrayFilter(x -> not match(x, '^\\\\d+$'), params) AS filtered_params
+            FROM metrics.pixels
+            WHERE (${pixelIDsWhereClause}) 
+            AND (${agentWhereClause})
+            AND request NOT ILIKE '%test=1%'
+            AND date = '${pastDate.toISOString().split('T')[0]}'
+            GROUP BY filtered_params;
+            `;
+
+        console.log('...Inserting data with query:');
+        console.log(queryString);
+
+        const clickhouseQuery = spawnSync('clickhouse-client', CH_ARGS.concat(queryString));
+        const resultErr = clickhouseQuery.stderr.toString();
+        if (resultErr) {
+            throw new Error(`Error creating table:\n ${resultErr}`);
+        }
+
+        pastDate.setDate(pastDate.getDate() + 1);
+    }
+    /* eslint-enable no-unmodified-loop-condition */
+}
+
+async function outputTableToCSV() {
+    console.log('Preparing CSV');
+
+    const chPromise = new Promise((resolve, reject) => {
+        const outputStream = fs.createWriteStream(PIXELS_TMP_CSV);
+        const queryString = `SELECT DISTINCT pixel, params FROM ${TMP_TABLE_NAME};`;
+        const clickhouseProcess = spawn('clickhouse-client', CH_ARGS.concat([queryString, '--format=CSVWithNames']));
+        clickhouseProcess.stdout.on('data', function (data) {
+            outputStream.write(data);
+        });
+
+        clickhouseProcess.stderr.on('data', function (data) {
+            reject(new Error(data.toString()));
+        });
+
+        clickhouseProcess.on('close', function (code) {
+            outputStream.end();
+            if (code !== 0) {
+                reject(new Error(`clickhouse-client process exited with code ${code}`));
+            }
+            resolve();
+        });
+    });
+
+    await Promise.all([chPromise])
+        .then(() => {
+            console.log('CSV file ready');
+        })
+        .catch((err) => {
+            console.error(err);
+            throw new Error('Error outputing data to CSV. Check logs above.');
+        });
+}
+
+function deleteTempTable() {
+    console.log('Deleting table');
+    const queryString = `DROP TABLE ${TMP_TABLE_NAME};`;
+    const clickhouseQuery = spawnSync('clickhouse-client', CH_ARGS.concat(queryString));
+    const resultErr = clickhouseQuery.stderr.toString();
+    if (resultErr) {
+        throw new Error(`Error deleting table:\n ${resultErr}`);
+    }
+}
+
+export async function preparePixelsCSV(mainPixelDir) {
+    createTempTable();
+    populateTempTable(getTokenizedPixels(mainPixelDir), getProductDef(mainPixelDir));
+    await outputTableToCSV();
+    deleteTempTable();
```

Comment:
How are we handling errors here? At the moment it looks like we'll just crash with the error, and the table data will remain. This will prevent the process from being rerun until the table has been removed. Should we either have:
 1. Some way of resuming a broken run.
 2. A cleanup option to the command to drop any existing table before starting.
 3. Just drop the table when we fail in an earlier step.

Reply:
Good catch. Went with 3